pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+	blockVolumePlugin, err :=
+		og.volumePluginMgr.FindMapperPluginByName(volumeToUnmount.PluginName)
+	if err != nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed", err)
+	}
+	if blockVolumePlugin == nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed to find BlockVolumeMapper plugin. Volume plugin is nil.", nil)
+	}
+	blockVolumeUnmapper, newUnmapperErr := blockVolumePlugin.NewBlockVolumeUnmapper(
+		volumeToUnmount.InnerVolumeSpecName, volumeToUnmount.PodUID)
+	if newUnmapperErr != nil {
+		return nil, blockVolumePlugin.GetPluginName(), volumeToUnmount.GenerateErrorDetailed("UnmapVolume.NewUnmapper failed", newUnmapperErr)
+	}
+
+	return func() error {
+		// Try to unmap symlink on global map path
+		globalUnmapPath, err :=
+			blockVolumeUnmapper.GetGlobalMapPath(volumeToUnmount.VolumeSpec)
+		if err != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToUnmount.GenerateErrorDetailed("UnmapVolume.GetGlobalUnmapPath failed", err)
+		}
+		unmapDeviceErr := og.blkUtil.UnmapDevice(globalUnmapPath, string(volumeToUnmount.PodUID))
+		if unmapDeviceErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToUnmount.GenerateErrorDetailed("UnmapVolume.UnmapDevice on global map path failed", unmapDeviceErr)
+		}
+
+		// Try to unmap symlink on pod device map path
+		podDeviceUnmapPath, volName := blockVolumeUnmapper.GetPodDeviceMapPath()
+		unmapDeviceErr = og.blkUtil.UnmapDevice(podDeviceUnmapPath, volName)
+		if unmapDeviceErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToUnmount.GenerateErrorDetailed("UnmapVolume.UnmapDevice on pod device map path failed", unmapDeviceErr)
+		}
+		removeLinkErr := og.blkUtil.RemoveMapPath(path.Join(podDeviceUnmapPath, volName))
+		if removeLinkErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToUnmount.GenerateErrorDetailed("UnmapVolume.RemoveSymlink failed", removeLinkErr)
+		}
+
+		glog.Infof(
+			"UnmapVolume succeeded for volume %q (OuterVolumeSpecName: %q) pod %q (UID: %q). InnerVolumeSpecName %q. PluginName %q, VolumeGidValue %q",
+			volumeToUnmount.VolumeName,
+			volumeToUnmount.OuterVolumeSpecName,
+			volumeToUnmount.PodName,
+			volumeToUnmount.PodUID,
+			volumeToUnmount.InnerVolumeSpecName,
+			volumeToUnmount.PluginName,
+			volumeToUnmount.VolumeGidValue)
+
+		// Update actual state of world
+		markVolUnmountedErr := actualStateOfWorld.MarkVolumeAsUnmounted(
+			volumeToUnmount.PodName, volumeToUnmount.VolumeName)
+		if markVolUnmountedErr != nil {
+			// On failure, just log and exit
+			glog.Errorf(volumeToUnmount.GenerateErrorDetailed("UnmapVolume.MarkVolumeAsUnmounted failed", markVolUnmountedErr).Error())
+		}
+
+		return nil
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapDeviceFunc(
```

Comment:
@jingxu97 can you help review this?  There's a lot cleanup steps going on here and we need to consider the cases where mounts could also be going on at the same time.

Reply:
For each volume, if operation_executor call this function with a unique podName, then two UnmapDevice operations from different pods could be executed concurrently. If you want to make sure for a single volume, only one UnmapDevice operation is executed at one time, you could use EmptyUniquePodName. See https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L677